javascript/ql/src/semmle/javascript/MembershipTests.qll

```diff
@@ -0,0 +1,257 @@
+    /**
+     * Gets the candidate of this test.
+     */
+    abstract DataFlow::Node getCandidate();
+
+    /**
+     * Gets a string that is a member of the collection of this test, if
+     * it can be determined.
+     */
+    string getAMemberString() { this.getAMemberNode().mayHaveStringValue(result) }
+
+    /**
+     * Gets a node that is a member of the collection of this test, if
+     * it can be determined.
+     */
+    DataFlow::Node getAMemberNode() { none() }
+
+    /**
+     * Gets the polarity of this test.
+     *
+     * If the polarity is `false` the test returns `true` if the
+     * collection does not contain the candidate.
+     */
+    boolean getPolarity() { result = true }
+  }
+}
+
+/**
+ * An `InclusionTest` viewed as a `MembershipTest`.
+ */
+class OrdinaryInclusionTest extends InclusionTest, MembershipTest::Range {
+  override DataFlow::Node getCandidate() { result = this.getContainedNode() }
+
+  override boolean getPolarity() { result = InclusionTest.super.getPolarity() }
+}
+
+/**
+ * A test for whether a candidate is a member of an array.
+ */
+class ArrayMembershipTest extends OrdinaryInclusionTest {
+  DataFlow::ArrayCreationNode array;
+
+  ArrayMembershipTest() { array.flowsTo(this.getContainerNode()) }
+
+  override DataFlow::Node getAMemberNode() { result = array.getAnElement() }
+}
+
+/**
+ * A test for whether a candidate is a member of an array constructed
+ * from a call to `String.prototype.split`.
+ */
+class ShorthandArrayMembershipTest extends OrdinaryInclusionTest {
+  StringSplitCall split;
+
+  ShorthandArrayMembershipTest() {
+    split.flowsTo(this.getContainerNode())
+  }
+
+  override string getAMemberString() {
+    exists(string toSplit |
+      split.getBaseString().mayHaveStringValue(toSplit) and
+      result = toSplit.splitAt(split.getSeparator())
+    )
+  }
+}
+
+/**
+ * An `EqualityTest` viewed as a `MembershipTest`.
+ */
+class EqualityLeftMembershipTest extends MembershipTest::Range, DataFlow::ValueNode {
+  override EqualityTest astNode;
+
+  override DataFlow::Node getCandidate() { astNode.getLeftOperand() = result.asExpr() }
+
+  override DataFlow::Node getAMemberNode() { result = astNode.getRightOperand().flow() }
+
+  override boolean getPolarity() { result = astNode.getPolarity() }
+}
+
+/**
+ * An `EqualityTest` viewed as a `MembershipTest`.
+ */
+class EqualityRightMembershipTest extends MembershipTest::Range, DataFlow::ValueNode {
+  override EqualityTest astNode;
+
+  override DataFlow::Node getCandidate() { astNode.getRightOperand() = result.asExpr() }
+
+  override DataFlow::Node getAMemberNode() { result = astNode.getLeftOperand().flow() }
+
+  override boolean getPolarity() { result = astNode.getPolarity() }
+}
+
+/**
+ * A switch case test viewed as a `MembershipTest`.
+ */
+class SwitchCaseMembershipTest extends MembershipTest::Range, DataFlow::ValueNode {
+  SwitchStmt switch;
+
+  SwitchCaseMembershipTest() { this = switch.getACase().getExpr().flow() }
```

Comment:
That is an excellent point. It does not look like we have any nodes for the implicit comparison that is performed at the `case`, so the choice of `this` here is arbitrary. 

To avoid feature creep, I have removed this class from the PR.

Looking forward: do we lean towards add a new synthetic data flow node for the implicit case comparisons, or towards complicating `MembershipTest` with an optional `getResultNode` (this won't solve the problem of where the test occurs)?